cpp/src/arrow/dataset/discovery_test.cc

```diff
@@ -364,6 +364,16 @@ TEST_F(FileSystemDatasetFactoryTest, FilenameNotPartOfPartitions) {
   }
 }
 
+TEST_F(FileSystemDatasetFactoryTest, UnparseablePartitionExpression) {
+  auto s = schema({field("first", int32()), field("second", int32())});
+  factory_options_.partitioning = std::make_shared<HivePartitioning>(s);
+
+  selector_.recursive = true;
+  MakeFactory({fs::File("first=one/file.parquet")});
+
+  ASSERT_RAISES(Invalid, factory_->Finish().status());
```

Comment:
Can you add some cases that shouldn't fail just in case, e.g.
```
""
"first=1/file.parquet"
"first=1/second=2/file.parquet"
```

Reply:
Will do